published/external/xdpapi.h

```diff
@@ -160,7 +232,269 @@ XdpUnloadApi(
     FreeLibrary(XdpHandle);
 }
 
-#endif // !defined(_KERNEL_MODE)
+#endif // defined(_KERNEL_MODE)
+
+#if defined(_KERNEL_MODE)
+
+#include <netioddk.h>
+#include <xdp/guid.h>
+
+#define POOLTAG_NMR_CLIENT      'cNdX'   // XdNc
+
+typedef struct _XDP_LOAD_CONTEXT *XDP_LOAD_API_CONTEXT;
+
+typedef
+VOID
+XDP_API_ATTACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef
+VOID
+XDP_API_DETACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef struct _XDP_API_CLIENT {
+    ERESOURCE Resource;
+
+    //
+    // NMR client registration.
+    //
+    NPI_CLIENT_CHARACTERISTICS NpiClientCharacteristics;
+    NPI_MODULEID ModuleId;
+    HANDLE NmrClientHandle;
+
+    //
+    // XDPAPI client.
+    //
+    VOID *Context;
+    XDP_API_ATTACH_FN *Attach;
+    XDP_API_DETACH_FN *Detach;
+    const XDP_API_CLIENT_DISPATCH *XdpApiClientDispatch;
+
+    //
+    // XDPAPI provider.
+    //
+    HANDLE BindingHandle;
+    XDP_API_PROVIDER_DISPATCH *XdpApiProviderDispatch;
+    XDP_API_PROVIDER_BINDING_CONTEXT *XdpApiProviderContext;
+} XDP_API_CLIENT;
+
+inline
+NTSTATUS
+XdpNmrClientAttachProvider(
+    _In_ HANDLE _NmrBindingHandle,
+    _In_ VOID *_ClientContext,
+    _In_ const NPI_REGISTRATION_INSTANCE *_ProviderRegistrationInstance
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_ClientContext;
+    NTSTATUS _Status;
+    VOID *_ProviderBindingContext;
+    const VOID *_ProviderBindingDispatch;
+
+    //
+    // The NMR client allows at most one active binding at a time, but defers
+    // all remaining binding logic to the NMR provider.
+    //
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&_Client->Resource);
+
+    if (_Client->BindingHandle != NULL) {
+        _Status = STATUS_DEVICE_NOT_READY;
+    } else if (_ProviderRegistrationInstance->Number != XDP_API_VERSION_1) {
+        _Status = STATUS_NOINTERFACE;
+    } else {
+        _Status =
+            NmrClientAttachProvider(
+                _NmrBindingHandle, _Client, &_Client->XdpApiClientDispatch,
+                &_ProviderBindingContext, &_ProviderBindingDispatch);
+
+        if (NT_SUCCESS(_Status)) {
+            _Client->BindingHandle = _NmrBindingHandle;
+            _Client->XdpApiProviderDispatch = (XDP_API_PROVIDER_DISPATCH *)_ProviderBindingDispatch;
+            _Client->XdpApiProviderContext = _ProviderBindingContext;
+        }
+    }
+
+    ExReleaseResourceAndLeaveCriticalRegion(&_Client->Resource);
+
+    if (NT_SUCCESS(_Status) && _Client->Attach != NULL) {
+        _Client->Attach(_Client->Context);
+    }
+
+    return _Status;
+}
+
+inline
+NTSTATUS
+XdpNmrClientDetachProvider(
```

Comment:
According to [MSDN](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nc-netioddk-npi_client_detach_provider_fn) this may be called at `DISPATCH_LEVEL`. This means we **must not** use locks that require `PASSIVE_LEVEL`, right?

Also, for the question of client cleaning up in the `Detach` call, that means it could not wait for cleanup, right? That seems weird, to require clean up always supporting `DISPATCH_LEVEL`...

Reply:
Also in the documentation, a client can return STATUS_PENDING and then call [NmrClientDetachProviderComplete](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nf-netioddk-nmrclientdetachprovidercomplete) at a later time. Client "cleanup" logic can be moved to [ClientCleanupBindingContext](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nc-netioddk-npi_client_cleanup_binding_context_fn) which happens after both client and provider have completed detach.

Reply:
As far as I understand, it is possible to use PASSIVE_LEVEL lock with DISPATCH_LEVEL function.
If multiple level components are mixed, lower level IRQL need to be used. PASSIVE_LEVEL.

Reply:
It is possible, but the code will be inherently buggy. NickB is correct that we must not use passive level locks. Windows Internals is a highly recommended resource for this concept, I'll send a link offline

Reply:
clientDetachProvider and providerDetachClient are called at dispatch level only when you deregister client/provider at dispatch level. If we have not treated these as functions that can be called at dispatch elsewhere in xdp like XdpNmrClientDetachProvider, why do we care now?

Reply:
> As far as I understand, it is possible to use PASSIVE_LEVEL lock with DISPATCH_LEVEL function.
As NickG mentioned, you cannot use a passive lock at DISPATCH_LEVEL (because the lock can block/wait, which isn't allowed at dispatch). But you can use a dispatch lock at PASSIVE_LEVEL.

> clientDetachProvider and providerDetachClient are called at dispatch level only when you deregister client/provider at dispatch level. If we have not treated these as functions that can be called at dispatch elsewhere in xdp like XdpNmrClientDetachProvider, why do we care now?

The NMR interface is generically defined to allow DISPATCH_LEVEL. My preference would be to not impose restrictions based on some assumptions for how XDP uses the interface today. What if XDP needs to be updated to deregister at dispatch in the future? @mtfriesen any opinion here?

Reply:
For code that is compiled into 3rd party code rather than something we own ourselves, we should do the right thing. IMO we should also fix code in XDP that makes incorrect assumptions about IRQL, but do that in a separate PR.